pkg/identity/identitymanager/manager.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identitymanager
+
+import (
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var (
+	// GlobalIdentityManager is a singleton instance of an IdentityManager, used
+	// for easy updating / tracking lifecycles of identities on the local node
+	// without having to pass around a specific instance of an IdentityManager
+	// throughout Cilium.
+	GlobalIdentityManager = NewIdentityManager()
+)
+
+// IdentityManager caches information about a set of identities, currently a
+// reference count of how many users there are for each identity.
+type IdentityManager struct {
+	mutex      lock.RWMutex
+	identities map[*identity.Identity]*identityMetadata
+}
+
+type identityMetadata struct {
+	refCount uint
+}
+
+// NewIdentityManager returns an initialized IdentityManager.
+func NewIdentityManager() *IdentityManager {
+	return &IdentityManager{
+		identities: make(map[*identity.Identity]*identityMetadata),
+	}
+}
+
+// Upsert inserts the identity into the GlobalIdentityManager.
+func Upsert(identity *identity.Identity) {
+	GlobalIdentityManager.Upsert(identity)
+}
+
+// Delete deletes the identity from the GlobalIdentityManager.
+func Delete(identity *identity.Identity) {
```

Comment:
To clarify, you're saying that I should change the names of the functions based on semantics as opposed to the underlying implementation?

Reply:
I was saying that both the implementation and naming should change to `Add()`/`Remove()`.

Semantically, if you use `Upsert()`, then you are telling the caller that they don't have to care about whether they insert the same identity twice. However, if the underlying implementation is doing reference tracking, then you want to enforce that the caller is appropriately adding and removing references to the identity for the duration that they are using the value. That way, this manager can know when no-one references the identity any more and can take action (eg, stop tracking the references and issue some event).

Reply:
I understand the distinction between changing it to `Add` and `Remove`, but I don't understand what you want me to change on the implementation side.

The `identitymanager` isn't aware of what is consuming the identities within it, just that something elsewhere is consuming the identities. 

Are you saying that we should also track what is consuming the identities within `identityManager`? E.g., a list of pointers to some interface, e.g., `Endpoint`, where `Endpoint` has a single function `GetSecurityIdentity`? This would ensure that we wouldn't be able to `Delete` if the consumer of the identity did not previously add it, as we would pass a pointer to the Endpoint to the `Add` and `Remove` functionality to indicate ownership.

I'm unsure how to otherwise enforce that the caller (the endpoint lifecycle) is appropriately adding and removing references to the identity for the duration that it is using the value apart from some type of integration test (which I am planning on adding once I get sign-off on the API).